CMakeLists.txt

```diff
@@ -94,15 +94,61 @@ target_sources(clio PRIVATE
 add_executable(clio_server src/main/main.cpp)
 target_link_libraries(clio_server PUBLIC clio)
 
+set(TEST_TARGET clio_tests)
+
 if(BUILD_TESTS)
-  add_executable(clio_tests
+    add_executable(${TEST_TARGET}
     unittests/RPCErrors.cpp
     unittests/Backend.cpp
     unittests/Logger.cpp
     unittests/Config.cpp
     unittests/ProfilerTest.cpp
     unittests/DOSGuard.cpp)
   include(CMake/deps/gtest.cmake)
+
+  # if CODE_COVERAGE enable, add clio_test-ccov
+  if(CODE_COVERAGE)
+      if("${CMAKE_C_COMPILER_ID}" MATCHES "(Apple)?[Cc]lang" OR "${CMAKE_CXX_COMPILER_ID}" MATCHES "(Apple)?[Cc]lang")
+          message("[Coverage] Building with llvm Code Coverage Tools")
+          # Using llvm gcov ; llvm install by xcode
+          set(LLVM_COV_PATH /Library/Developer/CommandLineTools/usr/bin)
+          if(NOT EXISTS ${LLVM_COV_PATH}/llvm-cov)
+              message(FATAL_ERROR "llvm-cov not found! Aborting.")
+          endif()
+
+          # set Flags
+          target_compile_options(${TEST_TARGET} PRIVATE -fprofile-instr-generate -fcoverage-mapping)
+          target_link_options(${TEST_TARGET} PRIVATE -fprofile-instr-generate -fcoverage-mapping)
+
+          # llvm-cov
+          add_custom_target(${TEST_TARGET}-ccov-preprocessing
+              COMMAND LLVM_PROFILE_FILE=${TEST_TARGET}.profraw $<TARGET_FILE:${TEST_TARGET}>
+              COMMAND ${LLVM_COV_PATH}/llvm-profdata merge -sparse ${TEST_TARGET}.profraw -o ${TEST_TARGET}.profdata
+              DEPENDS ${TEST_TARGET})
+
+          add_custom_target(${TEST_TARGET}-ccov-show
+              COMMAND ${LLVM_COV_PATH}/llvm-cov show $<TARGET_FILE:${TEST_TARGET}> -instr-profile=${TEST_TARGET}.profdata -show-line-counts-or-regions
+              DEPENDS ${TEST_TARGET}-ccov-preprocessing)
+
+          # add summary for CI parse
+          add_custom_target(${TEST_TARGET}-ccov-report
+              COMMAND ${LLVM_COV_PATH}/llvm-cov report $<TARGET_FILE:${TEST_TARGET}> -instr-profile=${TEST_TARGET}.profdata -ignore-filename-regex=".*_makefiles|.*unittests" -show-region-summary=false
+              DEPENDS ${TEST_TARGET}-ccov-preprocessing)
+
+          # exclude libs and unittests self
+          add_custom_target(${TEST_TARGET}-ccov
+              COMMAND ${LLVM_COV_PATH}/llvm-cov show $<TARGET_FILE:${TEST_TARGET}> -instr-profile=${TEST_TARGET}.profdata -show-line-counts-or-regions -output-dir=${TEST_TARGET}-llvm-cov -format="html" -ignore-filename-regex=".*_makefiles|.*unittests" > /dev/null 2>&1
+              DEPENDS ${TEST_TARGET}-ccov-preprocessing)
+
+          add_custom_command(TARGET ${TEST_TARGET}-ccov POST_BUILD
+              COMMENT "Open ${TEST_TARGET}-llvm-cov/index.html in your browser to view the coverage report."
```

Comment:
I think you can add something like `file://$(CMAKE_CURRENT_SOURCE_DIR)/${TEST_TARGET}-llvm-cov/index.html` so that it is a clickable link in the terminal? Not sure if `CMAKE_CURRENT_SOURCE_DIR` is a full path or you need some extra magic there.

Reply:
U 
![Screenshot 2023-01-16 at 11 25 29](https://user-images.githubusercontent.com/120398799/212672012-bb28a1eb-579c-4bd1-a78d-3099ec3430da.png)
can click it without extra work.